nats-base-client/request.ts

```diff
@@ -12,35 +12,132 @@
+import { RequestStrategy } from "./types.ts";
 import { ErrorCode, NatsError } from "./error.ts";
 import { MuxSubscription } from "./muxsubscription.ts";
 import { nuid } from "./nuid.ts";
 
-export class Request {
+export interface Request {
+  token: string;
+  requestSubject: string;
+  received: number;
+  resolver(err: Error | null, msg: Msg): void;
+  cancel(err?: NatsError): void;
+}
+
+export class BaseRequest {
   token: string;
   received: number;
-  deferred: Deferred<Msg>;
-  timer: Timeout<Msg>;
   ctx: Error;
   requestSubject: string;
-  private mux: MuxSubscription;
+  mux: MuxSubscription;
 
   constructor(
     mux: MuxSubscription,
     requestSubject: string,
-    opts: RequestOptions = { timeout: 1000 },
   ) {
     this.mux = mux;
     this.requestSubject = requestSubject;
     this.received = 0;
-    this.deferred = deferred();
     this.token = nuid.next();
-    extend(this, opts);
-    this.timer = timeout<Msg>(opts.timeout);
     this.ctx = new Error();
   }
+}
+
+export interface RequestManyOptionsInternal extends RequestManyOptions {
+  callback: (err: Error | null, msg: Msg | null) => void;
+}
+
+/**
+ * Request expects multiple message response
+ * the request ends when the timer expires,
+ * an error arrives or an expected count of messages
+ * arrives, end is signaled by a null message
+ */
+export class RequestMany extends BaseRequest implements Request {
+  callback!: (err: Error | null, msg: Msg | null) => void;
+  done: Deferred<void>;
+  timer: number;
+  max: number;
+  opts: Partial<RequestManyOptionsInternal>;
+  constructor(
+    mux: MuxSubscription,
+    requestSubject: string,
+    opts: Partial<RequestManyOptions> = { maxWait: 1000 },
+  ) {
+    super(mux, requestSubject);
+    this.opts = opts;
+    if (typeof this.opts.callback !== "function") {
+      throw new Error("callback is required");
+    }
+    this.callback = this.opts.callback;
+
+    this.max = typeof opts.maxMessages === "number" && opts.maxMessages > 0
+      ? opts.maxMessages
+      : -1;
+    this.done = deferred();
+    this.done.then(() => {
+      this.callback(null, null);
+    });
+    this.timer = setTimeout(() => {
+      this.cancel();
+    }, opts.maxWait);
+  }
+
+  cancel(err?: NatsError): void {
+    if (err) {
+      this.callback(err, null);
+    }
+    clearTimeout(this.timer);
+    this.mux.cancel(this);
+    this.done.resolve();
+  }
+
+  resolver(err: Error | null, msg: Msg): void {
+    if (err) {
+      err.stack += `\n\n${this.ctx.stack}`;
+      this.cancel(err as NatsError);
+    } else {
+      this.callback(null, msg);
+      if (this.opts.strategy === RequestStrategy.Count) {
+        this.max--;
+        if (this.max === 0) {
+          this.cancel();
+        }
+      }
+
+      if (this.opts.strategy === RequestStrategy.JitterTimer) {
+        clearTimeout(this.timer);
+        this.timer = setTimeout(() => {
+          this.cancel();
+        }, 300);
```

Comment:
If I understand correctly, with this mode, and if no "max count" is specified (I can't tell if this is mandatory or not, based on test you added, looks like it is not), then possibly you would get an unbounded number of replies, right? Meaning that although one could set a max_wait of 2sec, if a message comes, we reset the timer to fire in 300ms here, but I don't see something that would make it stop because we are passed the original max 2 sec...

Reply:
correct - this is the scatter pattern that Derek always talks about - I use it for ngs - Effectively it has an original deadline of some amount of time, but as messages come in that deadline is reset to a much smaller time. While the test was shrunk, I made do 100,000K+ messages, and I get all of them, possibly we'll want to expose that smaller timeout, but I figure that for now we are good.